### ISSUE TITLE
Represent Interfaces in the Fable AST

### ISSUE THREAD

**Issue by dbrattli** [fable-compiler/Fable#2673]
### Description

Interfaces are erased and not part of the Fable AST. For type annotations in languages such as Python and TypeScript it would be nice to be able to represent the interfaces as code. What is the current thinking about interfaces and e.g TypeScript? Would it be possible to add interfaces to the AST and let the language decide if it want's to erase them or not? I would be interested in contributing this, but want to check if there are any gotchas I should be aware of?

**Comment by dbrattli** [comment]
For example using type hints in Python I now produce code like this where the interface `AsyncObservable[T]` is erased to `Any`. This also reveals another problem as it's not allowed to use a generic parameter only once in Python (since it adds no information or value compared to Any).  I have a check that generic args are used more than once, but it's hard when the generic arguments ends up being erased. Anyways, the generated code is not really usable when types like `AsyncObservable[T]` becomes erased to `Any`. So it would be great if it was possible to represent interfaces as e.g abstract based classes in Python. 

<img width="1397" alt="Screenshot 2021-12-26 at 10 25 33" src="https://user-images.githubusercontent.com/849479/147404076-af08bf99-f322-472e-bc38-281bcb192091.png">
s

**Comment by dbrattli** [comment]
But I see now that there's not much info in the F# AST either:

```fs
type A =
    abstract member Method : arg: int -> int
```

Becomes (only):

```
[Entity (Tmp, [Entity (A, [])])]
```

So it might be hard to get this information?

**Comment by ncave** [comment]
@dbrattli Yes, the interface entity is there (in the AST) but unfortunately it doesn't actually have any of its members. Looks like the F# AST interface enitites only have their members when you get the interface from the actual object that implements it.

We tried for Rust a limited work-around to define the interface next to the first object that implements it, and replace all future references to that interface with the new namespace it was declared in, but that only works for future references, going forward, and Fable compiles files out of order, so it's not really a good solution.

A better solution will be to have a pre-compile step to scan all files and map all interface entities to their implementations.
It would be better if it was available in the F# AST, but I'm not sure that would be easy to add, so this is our best alternative, unless there is a better suggestion.

**Comment by alfonsogarciacaro** [comment]
@dbrattli @ncave Yes, sorry. This is actually something I wanted to do some time ago but forgot about it 😅 Adding interfaces to Fable AST, should be as simple as removing interfaces from `isIgnoredLeafEntity` here: https://github.com/fable-compiler/Fable/blob/030afc0dfd52d0a52e6679474893acc0fb61e4d4/src/Fable.Transforms/FSharp2Fable.fs#L1358-L1372

You don't get the interface members in the AST because the AST only contains implementations. But the important thing is to know where the interface is declared. With that and the `EntityRef`, it's easy to use `compiler.GetEntity(entityRef)` to get the Entity information and access `MembersFunctionsAndValues`. Let me try a quick PR.

**Comment by ncave** [comment]
@alfonsogarciacaro I may be wrong, but I'm not sure just removing interfaces from `isIgnoredLeafEntity` is enough, we need the interface members in order to define the interface in the right place, if it's declared in a different file than the implementation.
Hence my proposal (above) to (hopefully quickly) pre-process all entities in all files and generate a map of interfaces to implementation entities that have the interface members.

**Comment by alfonsogarciacaro** [comment]
@ncave Sorry, I may be misunderstanding you. If interfaces are included in the AST you should be able to just output the declaration when you encounter it, and for types implementing the interface you can just reference the interface as we [do with classes for reflection info](https://github.com/fable-compiler/Fable/blob/7db476d7897ae97fafffcb543f7a63b0b6125cba/src/Fable.Transforms/Fable2Babel.fs#L241), for example.

In any case, we [do build an entity map (per file) before compilation](https://github.com/fable-compiler/Fable/blob/7db476d7897ae97fafffcb543f7a63b0b6125cba/src/Fable.Transforms/State.fs#L87-L96), so it should be available if needed.

I sent a small PR that seems to work. It's just discarding the interfaces in Fable2Babel, but other languages (or typescript) could just access the information from `ent` [at this point](https://github.com/fable-compiler/Fable/pull/2675/files#diff-5c1d432616b2869d7a3e9828fcf99f13d2e4eb3e88d6abbd135f5cced457a628R2150-R2152) to declare the entity. Besides `MembersFunctionsAndValues` you can access `AllInterfaces` to see the inheriting interfaces, although it seems the collection also includes the current interface.

**Comment by dbrattli** [comment]
@alfonsogarciacaro This is awesome! I have tried the PR and I can confirm that I do find all the necessary information in `MembersFunctionsAndValues` ❤️

**Comment by ncave** [comment]
I stand corrected, it does seem to work on stand-alone interfaces. Thanks @alfonsogarciacaro!
